Docker Windows prerequisites & changes

MisterDA opened the pull request:
Some prerequisites for the Docker/Windows backend, and changes I gathered whilst working on Obuilder.

talex5 reviewed (commented):
There's a quite a mix of stuff here. I think it would be best to focus on what is needed for the Windows version. In particular, I suggest leaving the btrfs and zfs stuff alone if possible (unless it's some API change needed for Windows), and not reformatting the README.

MisterDA commented on lib/btrfs_store.ml:
```diff
@@ -1,8 +1,6 @@
-let strf = Printf.sprintf
-
-let running_as_root = Unix.getuid () = 0
+let str = Fmt.str
```
You're right, I completely overlooked this. I'm removing the commit.

MisterDA commented on lib/build.ml:
```diff
@@ -75,14 +75,14 @@ module Make (Raw_store : S.STORE) (Sandbox : S.SANDBOX) (Fetch : S.FETCHER) = st
         Lwt.finalize
           (fun () ->
-             cache |> Lwt_list.map_s (fun { Obuilder_spec.Cache.id; target; buildkit_options = _ } ->
+            cache |> Lwt_list.map_s (fun { Obuilder_spec.Cache.id; target; buildkit_options = _ } ->
```
Corrected, thanks.

MisterDA commented on lib/build.ml:
```diff
@@ -222,29 +223,29 @@ module Make (Raw_store : S.STORE) (Sandbox : S.SANDBOX) (Fetch : S.FETCHER) = st
-    >>!= fun id ->
-    let path = Option.get (Store.result t.store id) in
+    >>!= fun id -> Store.result t.store id
+    >>= fun path -> Lwt.return (Option.get path) >>= fun path ->
```
Monads, monads! fixed, thanks.

MisterDA commented on lib/build_log.ml:
```diff
@@ -74,6 +74,13 @@ let create path =
     len = 0;
   }
 
+let create_fd fd =
```
Applied, thanks.

MisterDA commented on lib/runc_sandbox.ml:
```diff
@@ -106,7 +103,7 @@ module Json_config = struct
     `Assoc fields
 
-  let make {Config.cwd; argv; hostname; user; env; mounts; network; mount_secrets} t ~config_dir ~results_dir : Yojson.Safe.t =
+  let make {Config.cwd; argv; hostname; user; env; mounts; network; mount_secrets; _} t ~config_dir ~results_dir : Yojson.Safe.t =
```
Turned it into an assertion that rejects an entrypoint for the runc sandbox.

MisterDA commented on lib/zfs_store.ml:
```diff
@@ -22,7 +22,7 @@ open Lwt.Infix
 *)
 
-let strf = Printf.sprintf
+let str = Fmt.str
```
As above, I made a mistake here. I've removed the commit.

MisterDA commented on stress/stress.ml:
```diff
@@ -21,38 +21,39 @@ module Fetcher = Docker
 
   let test_store t =
-    assert (Store.result t "unknown" = None);
+    Store.result t "unknown" >>= fun r -> assert (r = None); Lwt.return_unit >>= fun () ->
```
Thanks!

MisterDA commented on README.md:
```diff
@@ -1,81 +1,98 @@
 
-OBuilder takes a build script (similar to a Dockerfile) and performs the steps in it in a sandboxed environment.
+OBuilder takes a build script (similar to a Dockerfile) and performs
+the steps in it in a sandboxed environment.
```
I was introducing hard-breaks at 80 columns... like the old ones intended.
Removed the commit.

MisterDA commented:
Thank you for the review. I've removed all the commits that are not related to the Docker/Windows port, in particular the `Printf` to `Fmt` which was a gross mistake on my part. I've also addressed all your comments and ran ocp-indent on my changes.
Concerning the `Sandbox.run ?commit` commit, while crafting an answer describing why it was needed, I realized that the problem I was trying to solve -- I want to run jobs in the docker sandbox without "committing" (using the result of the run for the build) -- could be solved in the `build.ml` file, without the need for that parameter, so I've removed that commit too.

I'm starting to think that it was a bad idea to submit these prerequisites without the big picture, nevertheless your review helped me clean my code and my thought.

MisterDA commented:
Reading the other PRs, I realized that what I was trying to do with the _Allow writing build log to a file descriptor_ change is what #59 was trying to achieve: I want a file descriptor of the output of a sandbox run so that I can pipe it into another process. I was thinking of reusing the `Build_log` for that purpose, as the output doesn't need to be logged but just piped.

MisterDA commented:
Closing, will repush with all the commits.
